docs/proposals/cluster-autoscaling.md

```diff
@@ -0,0 +1,324 @@
+    Spec ClusterAutoscalerSpec `json:"spec,omitempty"`
+}
+```
+
+## Kubernetes 1.3 - medium KAC
+
+With medium KAC the goal is to replace all of the Google-specific parts of the autoscaling infrastructure
+with open-source code that provides the same functionality. This will make cluster autoscaling cloud-provider 
+independent (though an adapter layer will be needed for each cloud provider). It will take the required metrics from Heapster.
+
+
+```
+                          +--Nodes---------------------------+
+                          |                                  |
+                          | +--------+ +--------+ +--------+ |
+                          | |        | |        | |        | |
+                          | | Node1  | | Node2  | | NodeX  | |
+                          | |        | |        | |        | +<--+
+                          | +--------+ +--------+ +--------+ |   |
+            +---------+   | |cAdvisor| |cAdvisor| |cAdvisor| |   |
+            |  Config |   | +--------+ +--------+ +----+---+ |   |
+            +---------+   |     |          |           |     |   |
+                 |        +-----|----------|-----------|-----+   |
+                 |              |          |           |         |
+            +----v----+         |      +---v----+      |         |
+            |   KAC   |<-+      +----> |Heapster| <----+         |
+            +---------+  |             +---+----+                |
+                 |       |                 |                     |
+                 |       +-----------------+                     |
+ Kubernetes      |                                               |
+-----------------|-----------------------------------------------|------
+ Cloud provider  |                                               |
+                 |                                               |
+                 |                                       +-------+-------+
+                 +-------------------------------------> |Node controller|
+                                                         +---------------+
+```
+
+This will solve the following problems:
+(I2) [GCE specific logic] As cluster autoscaling will have most of its components inside the cluster it will be provider independent (I2).
+
+At this point we will encourage the community to contribute code that talks to particular cloud provider node controller.
+
+## Kubernetes 1.4-1.5 - big KAC
+
+Assuming the need for more sophisticated cluster autoscaling we will provide a full blown autoscaler that will solve (I3)[Works only for well formed pods] and will:
+
+* Take into account the number of pending pods and will check whether adding a node will allow to schedule them.
+* Choose the best node to kill once the downscale decision was made (and before will make sure that the homeless pods will schedule somewhere).
+* Reasonably handle batch-like best-effort workload
```

Comment:
I guess if you use request rather than utilization, this actually isn't a problem (especially if you assume batch requests 0).